nixos/modules/services/misc/matrix-appservice-discord.nix

```diff
@@ -0,0 +1,118 @@
+
+          Configuration options should match those described in
+          <link xlink:href="https://github.com/Half-Shot/matrix-appservice-discord/blob/master/config/config.sample.yaml">
+          config.sample.yaml</link>.
+
+          <option>config.bridge.domain</option> and <option>config.bridge.homeserverUrl</option>
+          should be set to match the public host name of the Matrix homeserver for webhooks and avatars to work.
+
+          Secret tokens should be specified using <option>environmentFile</option>
+          instead of this world-readable attribute set.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        example = ''
+          APPSERVICE_DISCORD_AUTH_CLIENT_I_D="12345"
+          APPSERVICE_DISCORD_AUTH_BOT_TOKEN="foobar"
+        '';
+        description = ''
+          File containing environment variables to be passed to the matrix-appservice-discord service,
+          in which secret tokens can be specified securely.
+        '';
+      };
+
+      homeserverService = mkOption {
+        type = types.string;
+        default = "matrix-synapse.service";
+        description = ''
+          Matrix homeserver service to wait for when starting the application service.
+        '';
+      };
+    };
+  };
+
+  config = let
+    cfg = config.services.matrix-appservice-discord;
+    appDir = "${pkgs.nodePackages.matrix-appservice-discord}/lib/node_modules/matrix-appservice-discord";
+    configJSON = pkgs.writeText "matrix-appservice-discord-config.json" (builtins.toJSON cfg.config);
+    configYAML = pkgs.runCommand "matrix-appservice-discord-config.yaml" { preferLocalBuild = true; } ''
+      ${pkgs.remarshal}/bin/json2yaml -i ${configJSON} -o $out
+    '';
+    registrationYAML = pkgs.runCommand "matrix-appservice-discord-registration.yaml" { preferLocalBuild = true; } ''
+      cd ${appDir}
+      ${pkgs.nodePackages.matrix-appservice-discord}/bin/matrix-appservice-discord \
+        --generate-registration \
+        --url='http://localhost:${toString cfg.port}' \
+        --config='${configYAML}' \
+        --file="$out"
+    '';
+  in mkIf cfg.enable {
+    systemd.services.matrix-appservice-discord = {
+      description = "A bridge between Matrix and Discord.";
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" cfg.homeserverService ];
+      after = [ "network-online.target" cfg.homeserverService ];
+      serviceConfig = {
+        Type = "simple";
+        WorkingDirectory = appDir;
+        ExecStart = ''
+          ${pkgs.nodePackages.matrix-appservice-discord}/bin/matrix-appservice-discord \
+            --file='${registrationYAML}' \
+            --config='${configYAML}' \
+            --port='${toString cfg.port}'
+        '';
+        Restart = "always";
+        DynamicUser = true;
+        StateDirectory = baseNameOf dataDir;
+        EnvironmentFile = cfg.environmentFile;
```

Comment:
When allowing a string and a path for the environment file we should also validate the file path in a `preStart` script.

Reply:
What are the criteria that we need to validate?

Reply:
I guess a simple `if [ ! -f "$file" ];` should be sufficient. This is just to ensure that a string path outside of the Nix store doesn't exist.

Reply:
Why would we want this file not to exist though?

Reply:
Just a mistake of mine, sorry ^^

Reply:
Also, `EnvironmentFile` already seems to ensure that the file exists.